### PR TITLE
CDI guide: add method removal documentation

### PR DESCRIPTION
**Pull request by loicmathieu**
Add to the CDI documentation the fact tha Arc also remove method that produces beans, not just beans, and that @Unremovable can be used on methods.

Also adds a line explaining how to debug removal issues

### CODE REVIEW

**Comment by geoand** [comment]
Maybe it's just me but I feel that this shouldn't be needed because a bean producer is a bean itself so what the documentation mentions still applies.
On the other hand, I understand that it can be useful to make things crystal clear.

**Comment by loicmathieu** [comment]
And this is needed, we had the case when working on the multiple mongoclient implementation with a Producer bean that is unremovable but not it's methods, and the methods was removed. We spent multiple hours searching what's happens then I discover by enabling the logs that the bean is not removed but it's methods are.

That's why I propose to add a sentence in the guide for this tricky case :)

**Comment by geoand** [comment]
Fair enough :)

**Comment by manovotn** [comment]
> a bean producer is a bean itself

That's right, producer method is equal declaration to a bean class.
And that's not something we came with either, that's from CDI spec.

Although we can of course make it clearer in our docs.

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 1]
```diff
@@ -474,9 +474,18 @@ Users can instruct the container to not remove any of their specific beans (even
...
 Furthermore, extensions can eliminate possible false positives by producing `UnremovableBeanBuildItem`.
 
+The container will also remove bean producing methods (methods annotated with `@Produces`) from producer beans if the method produces an unused bean.
+You can also annotate bean producing methods with `io.quarkus.arc.Unremovable`.
+
 Finally, Quarkus provides a middle ground for the bean removal optimization where application beans are never removed whether or not they are unused,
 while the optimization proceeds normally for non application classes. To use this mode, set `quarkus.arc.remove-unused-beans` to `fwk` or `framework`.
 
+If you feel that one of your beans, or one of your methods are removed,
+you can check it by enabling Arc producer beans logging via the following line in your application.properties.
```
Comment (manovotn) [inline docs/src/main/asciidoc/cdi-reference.adoc]: Let's reword this to:
`In order to see more information about which beans are being removed, you can enable additional logging via the following line in your application.properties.`

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 2]
```diff
@@ -474,9 +474,18 @@ Users can instruct the container to not remove any of their specific beans (even
  
 Furthermore, extensions can eliminate possible false positives by producing `UnremovableBeanBuildItem`.
 
+The container will also remove bean producing methods (methods annotated with `@Produces`) from producer beans if the method produces an unused bean.
+You can also annotate bean producing methods with `io.quarkus.arc.Unremovable`.
```
Comment (manovotn) [inline docs/src/main/asciidoc/cdi-reference.adoc]: I'd append this to the sentence above starting `Users can instruct...` in a following way:

```
Users can instruct the container to not remove any of their specific beans (even if they satisfy all the rules specified above) by annotating them with `io.quarkus.arc.Unremovable`.
This applies to all forms of bean declarations - bean class, producer method, producer field.
```

Would that be clear enough?

**Review by manovotn — COMMENTED** [review]
Added some suggestions.

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 1 above)
Comment (loicmathieu) [reply docs/src/main/asciidoc/cdi-reference.adoc]: done

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 2 above)
Comment (loicmathieu) [reply docs/src/main/asciidoc/cdi-reference.adoc]: I rephrase it a little to be crystal clear by splitting it to two sentences:

- One explaining that the optimization (removing beans) apply to all sort of bean definition
- One explaining that the annotation can be put on all sort of bean definition

I think that it worth being explicit in case the developer don't know the spec very well (like me more used to Spring DI than CDI)

**Comment by mkouba** [comment]
We should probably also clarify the `Unremovable`'s javadoc: "Indicates that the bean marked with this annotation should never be removed by Arc even if **it has no known injection points**". The bold part should be replaced with something like "it's considered unused" + add the description what "unused" means.

**Comment by loicmathieu** [comment]
@mkouba instead of copy/pasting the long description of what an unused bean is I propose to link to the corresponding paragraph in the guide:

_Indicates that the bean marked with this annotation should never be removed by Arc
 even if it's considered unused, you can find the description of an unused bean in the CDI reference guide: https://quarkus.io/guides/cdi-reference#remove_unused_beans_

File: docs/src/main/asciidoc/cdi-reference.adoc [hunk 3]
```diff
@@ -470,13 +470,23 @@ An unused bean:
 * does not declare any producer which is eligible for injection to any injection point,
 * is not directly eligible for injection into any `javax.enterprise.inject.Instance` or `javax.inject.Provider` injection point
 
+This optimization applies to all forms of bean declarations: bean class, producer method, producer field.
+
 Users can instruct the container to not remove any of their specific beans (even if they satisfy all the rules specified above) by annotating them with `io.quarkus.arc.Unremovable`.
+This annotation can be placed on the type declaration, on a producer method or on a producer field.
```
Comment (manovotn) [inline docs/src/main/asciidoc/cdi-reference.adoc]: The repetition of "on" in this sentence rings false in my ears. I am not a native speaker though; I could be wrong :)
How about we just say the following?
`This annotation can be placed on types, methods and fields.`

**Comment by manovotn** [comment]
> @mkouba instead of copy/pasting the long description of what an unused bean is I propose to link to the corresponding paragraph in the guide:

Link is fine by me, although having it all in javadoc is more comfortable as you don't need to leave the IDE just to read the javadoc elsewhere.

File: docs/src/main/asciidoc/cdi-reference.adoc (hunk 3 above)
Comment (loicmathieu) [reply docs/src/main/asciidoc/cdi-reference.adoc]: Yes but with the producer qualifier
```
This annotation can be placed on types, producer methods, and producer fields.
```

Comment (manovotn) [reply docs/src/main/asciidoc/cdi-reference.adoc]: Fair enough

Comment (loicmathieu) [reply docs/src/main/asciidoc/cdi-reference.adoc]: done

**Comment by loicmathieu** [comment]
> Link is fine by me, although having it all in javadoc is more comfortable as you don't need to leave the IDE just to read the javadoc elsewhere.

Re-think about it and I must agree :) 
I copy/paste the description from the guide in the JavaDoc

**Review by manovotn — APPROVED** [review]
